notifyme-sdk-backend/notifyme-sdk-backend-ws/src/main/java/ch/ubique/notifyme/sdk/backend/ws/util/CryptoUtilV3.java

```diff
@@ -0,0 +1,320 @@
+        }
+        try {
+            this.mpk = Hex.decodeHex(mpkHex);
+            this.mpkG2 = new G2();
+            this.mpkG2.deserialize(this.mpk);
+        } catch (DecoderException e) {
+            logger.error("unable to parse pk hexstring", e);
+            throw new RuntimeException(e);
+        }
+    }
+
+    public G2 getMpkG2() {
+        return mpkG2;
+    }
+
+    public void testFlow() {
+        Fr msk = new Fr();
+        msk.setByCSPRNG();
+        G2 mpk = new G2();
+        Mcl.mul(mpk, baseG2(), msk);
+
+        byte[] identity = createNonce(32);
+        G1 secretKeyForIdentity = keyDer(this.mskFr, identity);
+
+        // verifyTrace
+        int NONCE_LENGTH = 32;
+        byte[] msg_orig = createNonce(NONCE_LENGTH);
+        IBECiphertext ibeCiphertext = encryptInternal(this.mpkG2, identity, msg_orig);
+        byte[] msg_dec = decryptInternal(ibeCiphertext, secretKeyForIdentity, identity);
+        if (msg_dec == null) {
+            throw new RuntimeException("Health Authority could not verify Trace");
+        }
+    }
+
+    public List<TraceKey>
+    createTraceV3ForUserUpload(UserUploadPayloadOuterClass.UserUploadPayload userUpload) {
+        var traceKeys = new ArrayList<TraceKey>();
+        for (UserUploadPayloadOuterClass.UploadVenueInfo venueInfo : userUpload.getVenueInfosList()) {
+            if (!venueInfo.getFake()) {
+                byte[] identity =
+                        cryptoHashSHA256(
+                                concatenate(
+                                        "CN-ID".getBytes(StandardCharsets.US_ASCII),
+                                        venueInfo.getPreId().toByteArray(),
+                                        intToBytes(3600),
+                                        longToBytes(venueInfo.getIntervalStartMs() / 1000),
+                                        venueInfo.getTimeKey().toByteArray()));
+
+                G1 secretKeyForIdentity = keyDer(this.mskFr, identity);
+
+                byte[] nonce = createNonce();
+                NotifyMeAssociatedDataOuterClass.NotifyMeAssociatedData countryData =
+                        NotifyMeAssociatedDataOuterClass.NotifyMeAssociatedData.newBuilder()
+                                .setCriticality(NotifyMeAssociatedDataOuterClass.EventCriticality.LOW)
+                                .setVersion(1)
+                                .build();
+                byte[] encryptedAssociatedData =
+                        this.encryptAssociatedData(
+                                venueInfo.getNotificationKey().toByteArray(),
+                                "",
+                                countryData.toByteArray(),
+                                nonce,
+                                venueInfo.getIntervalStartMs(),
+                                venueInfo.getIntervalEndMs());
+
+                var traceKeyV3 = new ch.ubique.notifyme.sdk.backend.model.tracekey.v3.TraceKey();
+                traceKeyV3.setVersion(3);
```

Comment:
should this not be v4 now with the encrypted times?

Reply:
Resolved: See [https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/61d9e5717f2d092d2065c4058538a5d1db864a6b](https://github.com/SwissCovid/swisscovid-cn-backend/pull/2/commits/61d9e5717f2d092d2065c4058538a5d1db864a6b)